File: mc_pi.cpp
/* C++ program for estimation of Pi using Monte
   Carlo Simulation */
#include <bits/stdc++.h>
#include <random>
#include <cmath>
#include <chrono>

using namespace std::chrono;

int main(int argc,char* argv[])
{
    int i;
    double rand_x, rand_y, origin_dist, pi;
    double interval = 1e5;
    int circle_points = 0, square_points = 0;

    // START A TIMER FOR EXECUTION
    auto start = high_resolution_clock::now();

    // GENERATOR FOR RANDOM NUMBER UNIFORMLY DISTRIBUTED
    // IN (-1, 1)
    std::default_random_engine generator;
    std::uniform_real_distribution<double> distribution(-1,1);

    // ITERATION NUMBER IS TAKEN AS COMMANDLINE ARGUMENT (ONLY
    // IF PRESENT)
    if (argc > 1) interval =  atof(argv[1]);

    // LOOP FOR SPECIFIED NUMBER OF ITERATIONS
    for (i = 0; i <= interval; i++) {
        // Randomly generated x and y values
        rand_x = distribution(generator);
        rand_y = distribution(generator);

        // Distance between (x, y) from the origin
        origin_dist = rand_x * rand_x + rand_y * rand_y;

        // Checking if (x, y) lies inside the define
        // circle with R=1
        if (origin_dist <= 1) circle_points++;

        // Total number of points generated
        square_points++;

        // estimated pi after this iteration
        pi = double(4 * circle_points) / square_points;
    }

    // END TIME OF ROUTINE AND DURATION
    auto end = high_resolution_clock::now();
    duration<double> elapsed = end - start;

    // Final Estimated Value
    printf("\nAfter %.0e Iterations\n", interval);
    printf("\nEstimation of PI:\t PI = %f", pi);
    printf("\nCompare to CMATH:\t PI = %f\n", M_PI);
    std::cout << "\nExecution time was " <<
              elapsed.count() << "s.\n";

    return 0;
}

File: mc_pi.py
import time
import sys
import numpy as np
from numpy.random import default_rng


def pi_estimate(desired_iterations):
	"""
	Simple method of estimating PI by comparing randomly sampled point
	inside the area of a circle and a square (of diameter/length 1).
	
	:param desired_iterations: INT,
		Number of desired iterations to estimate PI
	:return: FLOAT,
		Estimate of PI
	"""
	rng = default_rng()
	
	increment = 0
	part_of_circle = 0
	part_of_square = 0
	while increment < desired_iterations:
		x = rng.uniform(-1.0, 1.0)
		y = rng.uniform(-1.0, 1.0)

		if x**2 + y**2 <= 1:
			part_of_circle += 1
	
		part_of_square += 1
		increment += 1
	
	return 4 * part_of_circle / part_of_square


def main():
	
	if len(sys.argv) > 1:
		iterations = float(sys.argv[1])
	else:
		iterations = 1e5
	
	start = time.time()
	monte_carlo_pi = pi_estimate(iterations)
	end = time.time()
	
	print(f"\nAfter {iterations:.0e} iterations\n\n"
	      f"Estimation of PI: \tPI = {monte_carlo_pi}\n"
	      f"Compare to NumPy: \tPI = {np.pi}\n\n"
	      f"Execution time was {end-start:f}s.\n")


if __name__ == "__main__":
	main()
